src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -390,21 +381,35 @@ public async Task UpdateBufferContentsAsync(CancellationToken cancellationToken)
                             edit.Replace(startPosition: 0, _textBuffer.CurrentSnapshot.Length, generatedSource.ToString());
                             edit.Apply();
                         }
+
+                        // If the file isn't already open, open it now. We may transition between opening and closing
+                        // if the file is repeatedly appearing and disappearing.
+                        var connectToWorkspace = _workspace.Options.GetOption(EnableOpeningInWorkspace) ?? false;
+
+                        if (connectToWorkspace && !_workspace.IsDocumentOpen(_documentIdentity.DocumentId))
+                        {
+                            _workspace.OnSourceGeneratedDocumentOpened(_documentIdentity, _textBuffer.AsTextContainer());
+                        }
                     }
                     finally
                     {
                         _updatingBuffer = false;
                     }
                 }
+                else
+                {
+                    // The file doesn't exist, so remove it from the workspace if it's still linked
+                    DisconnectFromWorkspaceIfOpen();
```

Comment:
can you expand on this comment more.  i don't know what it means. how can we be updating the file contents for soemthing that doesn't exist?

Reply:
these functions are getting so large and complex, it's very hard to reason about them at all.  i've spent several minutes and i can't even adequately tell what `if` block this `else` block is assocaited with.  We need to find ways to reign in this complexity.